### PR TITLE
Task #1276 - Rename templates to native Twig extension

### PR DESCRIPTION
**Pull request by andpavlenko**
Task #1276 - Templates renamed to native Twig extension .html.twig.

- Better understand code for new contributors.
- Syntax support for all IDEs without additional settings.

p.s. Some tuning PHP CS Fixer.

### CODE REVIEW

**Comment by evrifaessa** [comment]
this would really require us to bump the major version to 5.

**Comment by andpavlenko** [comment]
@evrifaessa Nothing breaking changes. Just convenience for future works.
Master branch accumulate changes for next major release. All as usual... ;)

**Comment by evrifaessa** [comment]
> @evrifaessa Nothing breaking changes. Just convenience for future works.
> Master branch accumulate changes for next major release. All as usual... ;)

It **is** a breaking change then. Basically anything that will break compability between two versions is a breaking change.

**Comment by andpavlenko** [comment]
@evrifaessa This is PR to master branch (not to release branch). Whether this major or minor release will be seen later, when we accumulated enough changes.

But if you would update stable release to master now - nothing will break (except private hardcoded changes).
Anyway: before update, users need to read [CHANGELOG.md](https://github.com/andpavlenko/boxbilling/blob/rename-templates/CHANGELOG.md#master) and all be fine. ;)

**Comment by andpavlenko** [comment]
Let's make major release after update admin theme.

**Comment by evrifaessa** [comment]
If this breaks the auto-update functionality, I wouldn't really like this to be merged without really measuring impact on users. No matter what, BoxBilling should be able to update itself, even for major releases. So we should either find a way to make the updater automatically update it, or leave it as is.

**Comment by andpavlenko** [comment]
@evrifaessa How does it break auto-update functionality?

**Comment by evrifaessa** [comment]
> @evrifaessa How does it break auto-update functionality?

it literally will create duplicates of the same template. the auto updater just downloads the prebuilt archive and extracts it over the existing files. it will not delete the old phtml files while extracting the new .twig ones. how are you going to handle this?

**Comment by givinghawk** [comment]
i have an idea,
you could make an update to the admin panel, that will show a button until the user presses it, that button would delete all of the old phtml files

**Comment by jaapmarcus** [comment]
Don't update over current auto update system... But force users to manually update and provide an upgrade path. 

You don't want users that currently use 7.4  or older to the new version any way as it will break any way due to lack of support. 

Any customisations will break any way so users still need to update manually and test their own *** 

You can't count on the fact that already run on PHP8.0 or higher, 
Don't have any custom plugins that will not break.

**Comment by andpavlenko** [comment]
> it literally will create duplicates of the same template

@evrifaessa How does it break functionality?

For those who would clean distribution:
1. Before upgrade need to read UPGRANDE.md
2. Don't need upgrade to master, only stable versions!
3. While tests passed - its mean all right! Or you need to make integration tests for checking structure of files. Since this is stupid - need to read UPGRANDE.md!

**This not only about templates! Look global:**
Impossible of upgrade file names, just does't allow migrate to PSR4, remove old and integrate convenient components! You never get rid of outdated Redbean and other hardcoded libraries... **Its make impossible any BB refactoring of code!**
Therefore for what [Looking for maintainers](https://github.com/boxbilling/boxbilling/issues/542)?

**Comment by evrifaessa** [comment]
I don't know if you really understood me because I'm having trouble understanding your message. Regular users should be able to upgrade painlessly right from the admin panel. 

Why would we want to break a working system and make it harder for people to upgrade? Who would want this? Stable versions will eventually have this patch merged too, what will happen then? Even when upgrading to another major release, the process should be seamless and with the least manual interference possible.

Yes, I absolutely agree with your point on finally ditching some old stuff and making use of newer standards and libraries, but we should also care about the people with little-to-no technical experience. If I was a user and I was told to migrate my templates manually, I'd seriously consider not using that software in the long run.

As our userbase grows bigger and bigger, we should start caring more about the impact of our PRs to our users.

TL;DR: I agree with your points but I don't want this to be merged without finding a way to upgrade seamlessly.

**Comment by andpavlenko** [comment]
> I don't know if you really understood me because I'm having trouble understanding your message.

_I try to write simple as possible._

If you don't understand what I'm talking about: please learn migration process of mature project (like Symfony 3->4, Drupal 6->7->8 and many other). You'll see only text documents for migration between major versions. Nobody make auto-upgrade file structures!

**Comment by andpavlenko** [comment]
> i have an idea, you could make an update to the admin panel, that will show a button until the user presses it, that button would delete all of the old phtml files

Best solution for weak users: manual remove *.phtml over FTP. Or more advanced:
```
cd project-dir
rm -rf ./*.phtml
```

But what about other files, which need to change during development? [Look global.](https://github.com/boxbilling/boxbilling/pull/1291#issuecomment-1107789342)

**Comment by BenNottelling** [comment]
> > i have an idea, you could make an update to the admin panel, that will show a button until the user presses it, that button would delete all of the old phtml files
> 
> Best solution for weak users: manual remove *.phtml over FTP. Or more advanced:
> 
> ```
> cd project-dir
> rm -rf ./*.phtml
> ```
> 
> But what about other files, which need to change during development? [Look global.](https://github.com/boxbilling/boxbilling/pull/1291#issuecomment-1107789342)

We have a PHP script that is run when you update.
It would not be a huge change to remove the leftover .phtml files from the themes in that file.

**Comment by andpavlenko** [comment]
> We have a PHP script that is run when you update.

Are you talking about the script `bin/update.sh`?

**Comment by jaapmarcus** [comment]
I assume he means via:
https://github.com/boxbilling/boxbilling/blob/76aee3bd68e045301db6d213a80a26b5e9486f89/src/bb-library/Box/Update.php

**Comment by andpavlenko** [comment]
> I assume he means via:

Need additional info.

p.s. I don't see any options yet than another bad practice. ((

**Comment by jaapmarcus** [comment]
public function performUpdate()

does preform the update  and unzips the folder … and does the thing… 

It is not ideal to implement a delete function for it …  

It is easier to force users upgrade via clean install

**Comment by andpavlenko** [comment]
> does preform the update and unzips the folder … and does the thing…

Any time when you need to change any file name you also need add custom migrate code in class???
1. Do you think this is reasonable?
2. How did you solve this problem before? I didn't find any similar migration in the code. Do you have example?
3. **This is make impossible migrate to install components via Composer!** Because auto-updater always will leave old vendor files in the system. @evrifaessa how will you solve this problem???

> It is easier to force users upgrade via clean install

Best idea! ;)
Just need save config and uploads.

**Comment by jaapmarcus** [comment]
> > does preform the update and unzips the folder … and does the thing…
> 
> Any time when you need to change any file name you also need add custom migrate code in class???

According this code yes... 

> 1. Do you think this is reasonable?
> 2. How did you solve this problem before? I didn't find any similar migration in the code. Do you have example?
> 3. **This is make impossible migrate to install components via Composer!** Because auto-updater always will leave old vendor files in the system. @evrifaessa how will you solve this problem???
> 

It makes sense how it was done in the past. But current methods has been improved ... 

> > It is easier to force users upgrade via clean install
> 
> Best idea! ;) Just need save config and uploads.

**Comment by andpavlenko** [comment]
> It makes sense how it was done in the past. But current methods has been improved ...

As I understand: If user use auto-updater, any run `composer update` **could leaves old files in the system** (auto-updater unzip new bb-vendor over old installed components)! And users even don't know about obsolete files! **Right?**

If yes, this is bigger problem then my templates rename which normally described in UPGRADE file!

I think current auto-updater make sense only for DB migrations and can't to be reliable method of CD.

**Comment by andpavlenko** [comment]
This PR contain improvement of PDF invoice design.

Screenshot - PDF Invoice

**Comment by evrifaessa** [comment]
> This PR contain improvement of PDF invoice design.
> 
This one looks awesome. Thanks for your work :)

**Comment by evrifaessa** [comment]
The screenshot looks nice! Looking forward to your work :)

**Comment by evrifaessa** [comment]
> Without identification of the application by anonymous user (for security reason).

I think we can discuss implementing further precautions for that. If we need such a thing, we can also move the "version" endpoint from the guest API interface to the admin API interface. We're planning to bump to a new major version, so I think we can be freer with the API changes.

**Comment by andpavlenko** [comment]
> > Without identification of the application by anonymous user (for security reason).
> 
> I think we can discuss implementing further precautions for that. If we need such a thing, we can also move the "version" endpoint from the guest API interface to the admin API interface. We're planning to bump to a new major version, so I think we can be freer with the API changes.

Please discuss about API changes not in this PR.

**Comment by andpavlenko** [comment]
Commit [07aafb1](https://github.com/boxbilling/boxbilling/pull/1291/commits/07aafb19fa0ed7eae853be8487daae777ca97a14) contains fully rewrite admin theme with Tabler.io. Still need to do the following:
- Improve dashboard
- Make some modal windows for small forms for simplify interface
- JS refactoring and remove jAlert in favor Bootstrap Modal
- Separate SVG from template and build SVG symbol sprite
- Cleanup outdated css/js files

Please carefully test on real data.

Screenshot from 2022-05-17 17-54-01

**Comment by evrifaessa** [comment]
I love where this is going. I think it will definitely look pleasing when it's ready to be merged. Thanks for all your work :)

**Comment by evrifaessa** [comment]
In my opinion, horizontal navbars (especially "navbar overlap") looks way better than the vertical ones. These are just my opinions though.

Navbar overlap:
Navbar overlap

Vertical:
Vertical

**Comment by evrifaessa** [comment]
Another thing, I think search bars are taking up too much space. We can move it to get rid of the big white gap or possibly make the search bar smaller.

Search bar

**Comment by evrifaessa** [comment]
`(...)/bb-admin/kb` -> Macro "table_search" is not defined in template "macro_functions.html.twig".

**Comment by evrifaessa** [comment]
Worth checking: https://preview.tabler.io/docs/flags.html

We're making use of flags in different parts of the software.

**Comment by andpavlenko** [comment]
> In my opinion, horizontal navbars (especially "navbar overlap") looks way better than the vertical ones. These are just my opinions though.

Its looks better with same designed widgets. But we can't make identical graphs widgets without massive code changes.

> Another thing, I think search bars are taking up too much space. We can move it to get rid of the big white gap or possibly make the search bar smaller.

Many pages contains identical code. Just different in advanced search filter:

- Clients (Overview/Advanced search)
- Orders (Overview/Advanced search)
- Invoices (Overview/Advanced search, Transaction overview/Transactions search)
- Support (Client tickets/Advanced ticket search)

I think, need to combine them in one page with slidable filter panel.
This will be make compact design and could to use advanced search without reload page.

> Worth checking: https://preview.tabler.io/docs/flags.html

Good idea. Need to try it.

**Comment by andpavlenko** [comment]
Align by center table look s better, I think.

Screenshot from 2022-05-19 19-14-19
